models/classes/accessControl/ActionAccessControl.php

```diff
@@ -0,0 +1,176 @@
+
+    /**
+     * @example [
+     *     'controller' => [
+     *         'action1' => [
+     *             'role1' => 'READ',
+     *             'role2' => 'WRITE,
+     *         ],
+     *     ],
+     * ]
+     */
+    public const OPTION_PERMISSIONS = 'permissions';
+
+    public const READ = 'READ';
+    public const WRITE = 'WRITE';
+    public const GRANT = 'GRANT';
+
+    /**
+     * @example $permissionsToAdd = [
+     *     'controller' => [
+     *         'action1' => [
+     *             'role1' => 'READ',
+     *             'role2' => 'WRITE,
+     *         ],
+     *     ],
+     * ]
+     */
+    public function addPermissions(array $permissionsToAdd = []): void
+    {
+        $permissions = $this->getOption(self::OPTION_PERMISSIONS, []);
+
+        foreach ($permissionsToAdd as $controller => $actions) {
+            if (empty($permissions[$controller])) {
+                $permissions[$controller] = $actions;
+
+                continue;
+            }
+
+            foreach ($actions as $action => $rolePermissions) {
+                if (empty($permissions[$controller][$action])) {
+                    $permissions[$controller][$action] = $rolePermissions;
+
+                    continue;
+                }
+
+                $actionPermissions = $permissions[$controller][$action] ?? [];
+                $permissions[$controller][$action] = array_merge($actionPermissions, $rolePermissions);
+            }
+        }
+
+        $this->setOption(self::OPTION_PERMISSIONS, $permissions);
+    }
+
+    /**
+     * @example $permissionsToRemove = [
+     *     'controller' => [
+     *         'action1' => [
+     *             'role1',
+     *             'role2',
+     *         ],
+     *     ],
+     * ]
+     */
+    public function removePermissions(array $permissionsToRemove = []): void
+    {
+        $permissions = $this->getOption(self::OPTION_PERMISSIONS, []);
+
+        foreach ($permissionsToRemove as $controller => $actions) {
+            foreach ($actions as $action => $roles) {
+                foreach ($roles as $role) {
+                    unset($permissions[$controller][$action][$role]);
+                }
+
+                if (empty($permissions[$controller][$action])) {
+                    unset($permissions[$controller][$action]);
+                }
+            }
+
+            if (empty($permissions[$controller])) {
+                unset($permissions[$controller]);
+            }
+        }
+
+        $this->setOption(self::OPTION_PERMISSIONS, $permissions);
+    }
+
+    public function hasReadAccess(string $controller, string $action, ?User $user = null): bool
+    {
+        return $this->hasAccess([self::READ, self::WRITE, self::GRANT], $controller, $action, $user);
+    }
+
+    public function hasWriteAccess(string $controller, string $action, ?User $user = null): bool
+    {
+        return $this->hasAccess([self::WRITE, self::GRANT], $controller, $action, $user);
```

Comment:
Should not be like this?

```suggestion
        return $this->hasAccess([self::WRITE], $controller, $action, $user);
```

Reply:
https://github.com/oat-sa/tao-core/pull/2883#discussion_r616088763

Reply:
Lets check together. I think there is some confusion here :)

Reply:
https://github.com/oat-sa/tao-core/pull/2883/files#r616375379